### DISCUSSION TITLE
Hibernate reactive panache pagination with OneToMany entity

### DISCUSSION THREAD

**Discussion by beiertu-mms** [quarkusio/quarkus#38643] [Q&A] [answered]
We're using [hibernate-reactive-panache v3.7.1](https://github.com/quarkusio/quarkus/tree/3.7.1/extensions/panache/hibernate-reactive-panache) with these two entities
```kotlin
@Entity
@Table(name = "owner")
@NamedQuery(name = "Owner.findById", query = "from Owner o left join fetch o._pets p where o.id = :id")
class Owner() {
    @OneToMany(
        cascade = [CascadeType.ALL],
        fetch = FetchType.LAZY,
        mappedBy = "owner",
        orphanRemoval = true
    )
    private val _pets = mutableListOf<Pet>()

    val pets get() = _pets.toList()

    fun addPet(pet: Pet) {
        if (_pets.contains(pet)) {
            throw IllegalArgumentException("The same pet can not be assigned to the same owner again!")
        }
        this._pets.add(pet)
    }
}

@Entity
@Table(name = "pet")
class Pet() {
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "owner_id")
    @JsonIgnore // ignore to prevent infinite recursion when serialization
    lateinit var owner: Owner
}
```

and we want to use paging as described [on the guide here](https://quarkus.io/guides/hibernate-orm-panache#paging).
But we're running into the [issue with join fetch](https://github.com/quarkusio/quarkus/issues/26308) when trying to get the paging information (pageCount, count, etc.) beside the fetch result list.

[This stackowerflow question](https://stackoverflow.com/questions/12459779/query-specified-join-fetching-but-the-owner-of-the-fetched-association-was-not) suggests to use a `countQuery` but that's not possible with the `NamedQuery` at the moment.

So we tried to add a second `NamedQuery` without the `join fetch` part
```kotlin
@NamedQuery(name = "Owner.countById", query = "from Owner o where o.id = :id")
```
to mimic the `countQuery` and tried to use both of them together like this

```kotlin
val params = Parameters.with("id", id)
ownerRepository
    .find("#Owner.findById", params).page<Owner>(pageIndex, pageSize).list<Owner>()
    .flatMap { owners ->
        val query = repo.find("#Owner.countById", params).page<Owner>(pageIndex, pageSize)
        Uni.combine().all().unis(
            Uni.createFrom().item(query.hasPreviousPage()),
            query.hasNextPage(),
            query.pageCount(),
            query.count()
        ).with { hasPrev, hasNext, pageCount, count ->
            OwnerPage(result = owners, hasPrev = hasPrev, hasNext = hasNext, pageCount = pageCount, count = count)
    }
}
```

but this throws the error `java.lang.IllegalStateException: Illegal pop() with non-matching JdbcValuesSourceProcessingState`.

Do you have any suggestions on how this can be implemented?

[a comment by an automated account is not shown]

**Comment by DavideD** [comment, marked as answer]
The problem is that, at the moment, operations using the session must run in a precise order. Basically, this means that you cannot use `Uni.combine().all().unis(...)`.
It should work if you compose/chain the operations.

**Reply by beiertu-mms** [reply]
yes, chaining the operations work :)
Thank you @DavideD
